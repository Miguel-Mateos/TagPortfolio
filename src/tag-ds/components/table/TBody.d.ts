import React from 'react';
export interface ITbodyProps {
    className?: string;
    [others: string]: any;
}
export declare const TBody: React.FC<ITbodyProps>;
