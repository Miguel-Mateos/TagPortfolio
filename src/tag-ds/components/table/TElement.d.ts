import React from 'react';
export interface ITableElement {
    className?: string;
    [others: string]: any;
}
export declare const TElement: React.FC<ITableElement>;
