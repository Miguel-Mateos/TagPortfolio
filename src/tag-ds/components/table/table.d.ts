import React from 'react';
export interface ITableProps {
    disabled?: boolean;
    className?: string;
    [others: string]: any;
}
export declare const TableWrapp: React.FC<ITableProps>;
export declare const Table: {
    Wrapper: React.FC<ITableProps>;
    Body: React.FC<import("./TBody").ITbodyProps>;
    Head: React.FC<import("./THead").ITHeadProps>;
    Element: React.FC<import("./TElement").ITableElement>;
    Row: React.FC<import("./TRow").ITRowProps>;
};
