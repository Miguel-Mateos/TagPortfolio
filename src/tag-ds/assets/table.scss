%table {
  width: 100%;
  @include font(medium);
  border-spacing: 0 calc(var(--unit) / 2);
  color: var(--neutral800);

  &[data-disabled='true'] tr *,
  &[data-disabled='true'] .td * {
    color: var(--neutral500);
    pointer-events: none;
  }

  & tr[data-disabled='true'] *,
  .tr[data-disabled='true'] * {
    color: var(--neutral500);
    pointer-events: none;
  }

  .table-menu-dropdown {
    width: auto;

    & .dropdown-menu {
      width: auto;
      transform: translateX(-50%);
    }
  }

  .table-menu-info {
    @extend %button-icon_interactive;
    width: calc(3 * var(--unit));
    height: calc(3 * var(--unit));
  }
}

%tbody {
  tr,
  .tr {
    border-radius: var(--unit);
    &:focus-visible {
      outline: calc(var(--unit) / 2) solid var(--secondary300);
    }

    &:hover:not(.active) {
      background-color: var(--neutral100);
    }

    & td:last-child,
    .td:last-child {
      border-radius: 0 var(--unit) var(--unit) 0;
    }

    & td:first-child,
    .td:first-child {
      border-radius: var(--unit) 0 0 var(--unit);
    }

    & td,
    .td {
      padding-left: 16px;
      .table-content {
        display: flex;
        // justify-content: center;
      }
    }

    & input[type='checkbox'] {
      margin: 0 !important;
    }

    // text-align: center;
    height: calc(8 * var(--unit));
  }

  & .active {
    background-color: var(--neutral200);
  }
}

%thead {
  tr,
  .tr {
    padding: 0 calc(2 * var(--unit));
    height: calc(6 * var(--unit));
    background-color: var(--neutral100);

    & td:last-child,
    .td:last-child {
      border-radius: 0 var(--unit) 0 0;
    }

    & td:first-child,
    .td:first-child {
      border-radius: var(--unit) 0 0 0;
    }

    & td,
    .td {
      padding-left: calc(2 * var(--unit));
      font-weight: 500;
      .table-content {
        display: flex;
        align-items: center;
      }
    }

    & input[type='checkbox'] {
      margin: 0 !important;
    }
  }

  &::after {
    content: '';
    height: var(--unit);
  }
}

.tag-ds {
  .table-wrapper,
  &.table-wrapper {
    width: 100%;
    overflow-x: auto;
  }
  table,
  .table,
  &.table {
    @extend %table;
  }

  .table-rotate-device {
    display: none;
  }

  tbody,
  .tbody,
  &.tbody {
    @extend %tbody;
  }

  thead,
  .thead,
  &.thead {
    @extend %thead;
  }

  @media (max-width: $mobile) {
    .table-wrapper {
      box-shadow: 0px calc(var(--unit) / 0.75) calc(3 * var(--unit)) rgba(28, 48, 75, 0.08);
      border-radius: var(--unit);
    }

    table,
    .table,
    &.table {
      border-spacing: unset;
    }

    thead {
      & tr > td:first-child {
        position: sticky;
        left: 0;
        background-color: var(--neutral100);
        border-right: 1px solid var(--neutral200);
      }
    }

    tbody {
      tr {
        &:hover:not(.active) td:first-child {
          background-color: var(--neutral100);
        }
        &.active {
          & td:first-child {
            background-color: var(--neutral200);
          }
        }
      }

      td {
        min-width: 160px;
      }

      &.table-rotate-device,
      .table-rotate-device {
        color: var(--neutral700);
        display: flex;
        & .material-icons {
          font-size: var(--fontSizeMedium);
          font-weight: 400;
          line-height: var(--lineHeightMedium);
          margin-right: var(--unit);
        }
        & > .table-rotate-device-text {
          font-size: var(--fontSizeMedium);
          font-weight: 400;
          line-height: var(--lineHeightMedium);
        }
      }
    }

    tbody,
    .tbody,
    &.tbody {
      & tr > td:first-child {
        position: sticky;
        left: 0;
        background-color: var(--neutral0);
        border-right: 1px solid var(--neutral200);
        border-radius: 0;
      }
      & tr > td:last-child {
        border-radius: 0;
      }
    }
  }

  @media (max-width: $mobile) and (orientation: landscape) {
    .table-wrapper {
      box-shadow: 0px calc(var(--unit) / 0.75) calc(3 * var(--unit)) rgba(28, 48, 75, 0.08);
      border-radius: var(--unit);
    }

    .table-rotate-device {
      display: none;
    }
  }
}

table.tag-ds {
  @extend %table;

  .table-rotate-device {
    display: none;
  }

  tbody,
  .tbody,
  &.tbody {
    @extend %tbody;
  }

  thead,
  .thead,
  &.thead {
    @extend %thead;
  }

  @media (max-width: $mobile) {
    .table-wrapper {
      box-shadow: 0px calc(var(--unit) / 0.75) calc(3 * var(--unit)) rgba(28, 48, 75, 0.08);
      border-radius: var(--unit);
    }

    table,
    .table,
    &.table {
      border-spacing: unset;
    }

    thead {
      & tr > td:first-child {
        position: sticky;
        left: 0;
        background-color: var(--neutral100);
        border-right: 1px solid var(--neutral200);
      }
    }

    tbody {
      tr {
        &:hover:not(.active) td:first-child {
          background-color: var(--neutral100);
        }
        &.active {
          & td:first-child {
            background-color: var(--neutral200);
          }
        }
      }

      td {
        min-width: 160px;
      }

      &.table-rotate-device,
      .table-rotate-device {
        color: var(--neutral700);
        display: flex;
        & .material-icons {
          font-size: var(--fontSizeMedium);
          font-weight: 400;
          line-height: var(--lineHeightMedium);
          margin-right: var(--unit);
        }
        & > .table-rotate-device-text {
          font-size: var(--fontSizeMedium);
          font-weight: 400;
          line-height: var(--lineHeightMedium);
        }
      }
    }

    tbody,
    .tbody,
    &.tbody {
      & tr > td:first-child {
        position: sticky;
        left: 0;
        background-color: var(--neutral0);
        border-right: 1px solid var(--neutral200);
        border-radius: 0;
      }
      & tr > td:last-child {
        border-radius: 0;
      }
    }
  }

  @media (max-width: $mobile) and (orientation: landscape) {
    .table-wrapper {
      box-shadow: 0px calc(var(--unit) / 0.75) calc(3 * var(--unit)) rgba(28, 48, 75, 0.08);
      border-radius: var(--unit);
    }

    .table-rotate-device {
      display: none;
    }
  }
}

