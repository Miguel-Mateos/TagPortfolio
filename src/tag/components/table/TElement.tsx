import React from 'react';

export interface ITableElement {
  className?: string
  [others: string]: any
}

export const TElement: React.FC<ITableElement> = ({ children, className, ...rest }) => (
  <td className={className || ''}  {...rest}><div className='table-content'>{children}</div></td>
);