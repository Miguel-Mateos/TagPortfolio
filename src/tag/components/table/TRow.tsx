import React from 'react';

export interface ITRowProps {
  active?: boolean
  disabled?: boolean
  className?: string
  [others: string]: any
}
 
export const TRow: React.FC<ITRowProps> = ({children, active, disabled, className, ...rest }) => { 
  return(
    <tr data-disabled={disabled} className={`${className || ''}${active ? 'active' : ''}`} {...rest}>{children}</tr>
  );
};