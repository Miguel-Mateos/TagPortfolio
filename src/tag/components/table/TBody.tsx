import React, { useCallback, useState } from 'react';
import { TRow } from './TRow';

export interface ITbodyProps {
  className?: string
  [others: string]: any
}

export const TBody: React.FC<ITbodyProps> = ({ children, className, ...rest }) => { 
  const [active, setActive] = useState<number|null>(null);
  const isValid = useCallback(() => {
    return React.Children.toArray(children).every((_child: any) => _child.type === (TRow || 'tr'));
  }, [children]);

  const overpoweredChildren = React.Children.toArray(children).map((_child: any, idx) => {
    return React.cloneElement((_child), {
      'data-testid': rest['data-testid'] ? rest['data-testid']+'-body-row-'+idx : undefined,
      active: idx === active,
      onClick: () => idx === active ? setActive(null) : setActive(idx)
    });
  });

  if (isValid()) return (
    <tbody className={className || 'rest'} {...rest}>{overpoweredChildren}</tbody>
  );

  return null;
};