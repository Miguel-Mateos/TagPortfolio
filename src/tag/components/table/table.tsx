import React from 'react';
import { TBody } from './TBody';
import { TElement } from './TElement';
import { THead } from './THead';
import { TRow } from './TRow';
export interface ITableProps {
  disabled?: boolean
  className?: string
  [others: string]: any
}

export const TableWrapp: React.FC<ITableProps> = ({ disabled, children, className, ...rest }) => {

  return (
    <>
      <div className='table-rotate-device'>
        <span className="material-icons">screen_rotation</span>
        <span className='table-rotate-device-text'>Rotate your device</span>
      </div>
      <div className='table-wrapper'>
        <table data-disabled={disabled} className={className || ''}
          {...rest}
          data-testid={rest['data-testid'] ? rest['data-testid']+'-wrapper' : ''}>
          {children}
        </table>
      </div>
    </>
  );
};

export const  Table = {
  Wrapper: TableWrapp,
  Body: TBody,
  Head: THead,
  Element: TElement,
  Row: TRow 
};
