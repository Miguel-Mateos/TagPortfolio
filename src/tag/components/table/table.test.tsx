import React from 'react';
import { Table, ITableProps } from './table';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const leftElements = [
  <div
    key={1}
    className='checkbox-container'
  >
    <input id="checkbox" onChange={(e) => console.log(0, e)} type="checkbox" />
  </div>,

];

const rightElements = [
  <div
    key={1}
    className='checkbox-container'
  >
    <input id="checkbox" onChange={(e) => console.log(0, e)} type="checkbox" />
  </div>,
];

const elements = [
  [ 'Content',
    'Content',
    'Content',
    'Content',
    'Content',
    'Content',
    'Content'
  ],
  [ 'Content',
    'Content',
    'Content',
    'Content',
    'Content',
    'Content',
    'Content'
  ],
  [ 'Content',
    'Content',
    'Content',
    'Content',
    'Content',
    'Content',
    'Content'
  ]
];

const headElements = [
  'Label',
  'Label',
  'Label',
  'Label',
  'Label',
  'Label',
  'Label',
];

const TableTest = (args: ITableProps) => {
  return <div className="tag-ds App" style={{ height: '30vh', width: '50vw'}}>
    <Table.Wrapper {...args}>
      <Table.Head>
        <Table.Row>
          {leftElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          {headElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          {rightElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
        </Table.Row>
      </Table.Head>
      <Table.Body data-testid="table">
        {elements.map((_,idx) => (
          <Table.Row key={idx}>
            {leftElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
            {_.map(elem => <Table.Element key={idx+'string'}>{elem}</Table.Element>)}
            {rightElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          </Table.Row>
        ))}
      </Table.Body>
    </Table.Wrapper>
  </div>;
};

const InvalidTableTest = (args: ITableProps) => {
  return <div className="tag-ds App" style={{ height: '30vh', width: '50vw'}}>
    <Table.Wrapper {...args}>
      <Table.Head>
        <Table.Row>
          {leftElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          {headElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          {rightElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
        </Table.Row>
      </Table.Head>
      <Table.Body data-testid="table">
        {elements.map((_,idx) => (
          <div key={idx}>
            {leftElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
            {_.map(elem => <Table.Element key={idx+'string'}>{elem}</Table.Element>)}
            {rightElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          </div>
        ))}
      </Table.Body>
    </Table.Wrapper>
  </div>;
};

const TheadTableTest = (args: ITableProps) => {
  return <div className="tag-ds App" style={{ height: '30vh', width: '50vw'}}>
    <Table.Wrapper {...args}>
      <Table.Head>
        <Table.Row>
          {leftElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          {headElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          {rightElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
        </Table.Row>
      </Table.Head>
      <Table.Body data-testid="table">
        {elements.map((_,idx) => (
          <tr key={idx}>
            {leftElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
            {_.map(elem => <Table.Element key={idx+'string'}>{elem}</Table.Element>)}
            {rightElements.map(elem => <Table.Element key={'test'}>{elem}</Table.Element>)}
          </tr>
        ))}
      </Table.Body>
    </Table.Wrapper>
  </div>;
};


test('renders and matches snapshot', () => {
  const component = renderer.create(<TableTest/>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders and click on row', () => {
  const { getByTestId } = render(<TableTest data-testid='table'/>);
  const row = getByTestId('table-body-row-1');
  userEvent.click(row);
  expect(row).toHaveClass('active');
});

test('render disabled table', () => {
  const { getByTestId } = render(<TableTest disabled={true} data-testid='table'/>);
  const disabledTable = getByTestId('table-wrapper');
  expect(disabledTable).toHaveAttribute('data-disabled', 'true');
});

test('render with invalid body', () => {
  const { getByTestId } = render(<InvalidTableTest disabled={true} data-testid='table'/>);
  const disabledTable = getByTestId('table-wrapper');
  expect(disabledTable).toHaveAttribute('data-disabled', 'true');
});

test('render with invalid thead', () => {
  const { getByTestId } = render(<TheadTableTest disabled={true} data-testid='table'/>);
  const disabledTable = getByTestId('table-wrapper');
  expect(disabledTable).toHaveAttribute('data-disabled', 'true');
});