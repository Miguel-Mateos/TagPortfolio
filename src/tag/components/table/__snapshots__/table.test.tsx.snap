// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot 1`] = `
<div
  className="tag-ds App"
  style={
    Object {
      "height": "30vh",
      "width": "50vw",
    }
  }
>
  <div
    className="table-rotate-device"
  >
    <span
      className="material-icons"
    >
      screen_rotation
    </span>
    <span
      className="table-rotate-device-text"
    >
      Rotate your device
    </span>
  </div>
  <div
    className="table-wrapper"
  >
    <table
      className=""
      data-testid=""
    >
      <thead
        className=""
      >
        <tr
          className=""
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Label
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
      </thead>
      <tbody
        className="rest"
        data-testid="table"
      >
        <tr
          className=""
          data-testid="table-body-row-0"
          onClick={[Function]}
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
        <tr
          className=""
          data-testid="table-body-row-1"
          onClick={[Function]}
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
        <tr
          className=""
          data-testid="table-body-row-2"
          onClick={[Function]}
        >
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              Content
            </div>
          </td>
          <td
            className=""
          >
            <div
              className="table-content"
            >
              <div
                className="checkbox-container"
              >
                <input
                  id="checkbox"
                  onChange={[Function]}
                  type="checkbox"
                />
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
